crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -153,70 +154,105 @@ where
+                .into_iter()
+                .map(|seq_num| {
+                    self.download_checkpoint_data(seq_num as u64)
+                });
+            let download_results = join_all(download_futures).await;
+            let mut downloaded_checkpoints = vec![];
+            // NOTE: Push sequentially and if one of the downloads failed,
+            // we will discard all following checkpoints and retry, to avoid messing up the DB commit order.
+            for download_result in download_results {
+                if let Ok(checkpoint) = download_result {
+                    downloaded_checkpoints.push(checkpoint);
+                } else {
+                    break;
+                }
+            }
+
+            next_cursor_sequence_number += downloaded_checkpoints.len() as i64;
+            // NOTE: with this line, we can make sure that:
+            // - when indexer is way behind and catching up, we download MAX_PARALLEL_DOWNLOADS checkpoints in parallel;
+            // - when indexer is up to date, we download at least one checkpoint at a time.
+            current_parallel_downloads =
+                std::cmp::min(downloaded_checkpoints.len() + 1, MAX_PARALLEL_DOWNLOADS);
+            if downloaded_checkpoints.is_empty() {
+                continue;
+            }
+
+            // Index checkpoint data
+            let index_guard = self.metrics.checkpoint_index_latency.start_timer();
+            let indexed_checkpoint_epoch_vec = downloaded_checkpoints
+                .iter()
+                .map(|downloaded_checkpoint| self.index_checkpoint(&downloaded_checkpoint))
+                .collect::<Result<Vec<_>, IndexerError>>()
+                .map_err(|e| {
                     error!(
-                        "Failed to download checkpoint data with checkpoint sequence number {} and error {:?}, retrying...",
-                        next_cursor_sequence_number, e
+                        "Failed to index checkpoints {:?} with error: {}",
+                        downloaded_checkpoints,
+                        e.to_string()
                     );
                     e
                 })?;
-            self.metrics.total_checkpoint_received.inc();
-
-            // Index checkpoint data
-            let index_guard = self.metrics.checkpoint_index_latency.start_timer();
-            let (indexed_checkpoint, indexed_epoch) = self.index_checkpoint(&checkpoint)?;
+            let (indexed_checkpoints, indexed_epochs): (Vec<_>, Vec<_>) =
+                indexed_checkpoint_epoch_vec.into_iter().unzip();
             index_guard.stop_and_record();
 
             let checkpoint_sender_guard = self.checkpoint_sender.lock().await;
             // NOTE: when the channel is full, checkpoint_sender_guard will wait until the channel has space.
-            checkpoint_sender_guard
+            // Also add new checkpoint sequentially to stick to checkpoint order.
+            for indexed_checkpoint in indexed_checkpoints {
```

Comment:
There is still room for improvement here. Right now, the download of batch n+1 is blocked by the committing of batch n(i.e., downloading and db committing are not pipelined). The approach mentioned [here](https://github.com/MystenLabs/sui/pull/10408#discussion_r1157989271) can further remove this overhead